File: frontend/src/lib/weather.ts
// Погода на стартах — общие типы и форматирование для протокола, пробежек,
// локации и обзора. Значения и подпись (label/icon/summary) приходят с бэкенда
// (start_weather_service), здесь — только вывод.

export type WeatherBrief = {
  date: string;
  start_time_local: string | null;
  temperature_c: number | null;
  apparent_temperature_c: number | null;
  humidity_pct: number | null;
  weather_code: number | null;
  label: string;
  icon: string;
  wind_speed_ms: number | null;
  wind_gusts_ms: number | null;
  /** Осадки в окне «час до старта — два часа после», мм. */
  precipitation_run_mm: number | null;
  precipitation_before_mm: number | null;
  snowfall_cm: number | null;
  snow_depth_cm: number | null;
  day_temperature_min_c: number | null;
  day_temperature_max_c: number | null;
  day_precipitation_mm: number | null;
  sunrise_local: string | null;
  rain_kind: "dry" | "drizzle" | "rain" | "downpour";
  is_rain: boolean;
  is_snow_cover: boolean;
  /** Предварительная строка из прогнозной модели (субботний вечер), архив заменит. */
  is_preliminary: boolean;
  summary: string;
};

/** Прогноз на ближайший старт: цифры, готовая строка и советы. */
export type WeatherForecast = {
  target_date: string;
  start_time_local: string | null;
  temperature_c: number | null;
  apparent_temperature_c: number | null;
  humidity_pct: number | null;
  precipitation_mm: number | null;
  precipitation_probability_pct: number | null;
  snowfall_cm: number | null;
  weather_code: number | null;
  label: string;
  icon: string;
  cloud_cover_pct: number | null;
  wind_speed_ms: number | null;
  wind_gusts_ms: number | null;
  /** За сколько суток до старта снят прогноз и как это читать. */
  horizon_days: number;
  horizon_note: string;
  updated_at: string | null;
  summary: string;
  advice: string[];
};

export type WeatherStartRef = {
  platform_code: string;
  event_number: number | null;
  finishers: number | null;
};

export type WeatherRecord = { weather: WeatherBrief; start: WeatherStartRef | null };

/** «−27°» / «12°» — целые градусы, настоящий минус. */
export function formatTemp(value: number | null | undefined): string {
  if (value == null || Number.isNaN(value)) {
    return "—";
  }
  // Половинки — от нуля (−43.5 → −44), как в format_temperature на бэкенде:
  // иначе плитка и summary одной пробежки расходились бы на градус.
  const rounded = Math.sign(value) * Math.round(Math.abs(value));
  return rounded < 0 ? `−${Math.abs(rounded)}°` : `${Math.abs(rounded)}°`;
}

/** Короткая подпись для таблиц: значок + градусы («🌧️ 12°»). */
export function weatherShort(weather: WeatherBrief | null | undefined): string {
  if (!weather) {
    return "—";
  }
  return `${weather.icon} ${formatTemp(weather.temperature_c)}`.trim();
}

/** Подсказка к короткой подписи: полная строка + пометка о предварительности. */
export function weatherTitle(weather: WeatherBrief | null | undefined): string | undefined {
  if (!weather) {
    return undefined;
  }
  return weather.is_preliminary
    ? `${weather.summary} · предварительно, архив уточнит в понедельник`
    : weather.summary;
}

export const RAIN_KIND_LABELS: Record<WeatherBrief["rain_kind"], string> = {
  dry: "без дождя",
  drizzle: "возможно, моросило",
  rain: "дождь",
  downpour: "ливень",
};

/** Класс тона для температуры: мороз — синий, жара — красный, остальное нейтрально. */
export function temperatureTone(value: number | null | undefined): "frost" | "cold" | "mild" | "warm" | "hot" | "none" {
  if (value == null) {
    return "none";
  }
  if (value <= -10) {
    return "frost";
  }
  if (value < 0) {
    return "cold";
  }
  if (value < 20) {
    return "mild";
  }
  if (value < 28) {
    return "warm";
  }
  return "hot";
}

File: backend/app/schemas/weather.py
"""Погода на стартах — ответы API (см. start_weather_service)."""

from __future__ import annotations

from pydantic import BaseModel, Field


class WeatherBriefResponse(BaseModel):
    """Погода в час старта одной даты + готовая строка summary."""

    date: str
    start_time_local: str | None = None
    temperature_c: float | None = None
    apparent_temperature_c: float | None = None
    humidity_pct: int | None = None
    weather_code: int | None = None
    label: str = ""
    icon: str = ""
    wind_speed_ms: float | None = None
    wind_gusts_ms: float | None = None
    # Осадки в окне «старт−1ч…старт+2ч» и «старт−4ч…старт−1ч» (трасса мокрая).
    precipitation_run_mm: float | None = None
    precipitation_before_mm: float | None = None
    snowfall_cm: float | None = None
    snow_depth_cm: float | None = None
    day_temperature_min_c: float | None = None
    day_temperature_max_c: float | None = None
    day_precipitation_mm: float | None = None
    sunrise_local: str | None = None
    # dry / drizzle / rain / downpour — см. пороги в start_weather_service.
    rain_kind: str = "dry"
    is_rain: bool = False
    is_snow_cover: bool = False
    # Предварительная строка из прогнозной модели (субботний вечер), архив заменит.
    is_preliminary: bool = False
    summary: str


class WeatherStartRefResponse(BaseModel):
    platform_code: str
    event_number: int | None = None
    finishers: int | None = None


class WeatherRecordResponse(BaseModel):
    weather: WeatherBriefResponse
    start: WeatherStartRefResponse | None = None


class LocationWeatherMonthResponse(BaseModel):
    month: int
    label: str
    # Сколько стартов площадки пришлось на этот месяц (с погодой).
    starts: int = 0
    # starts — статистика по стартам; saturdays — стартов в месяце не было, взяты все субботы; none — данных нет.
    basis: str = "none"
    samples: int = 0
    temperature_median_c: float | None = None
    temperature_min_c: float | None = None
    temperature_max_c: float | None = None
    apparent_median_c: float | None = None
    rain_share: float | None = None
    snow_share: float | None = None


class LocationWeatherAttendanceResponse(BaseModel):
    key: str
    label: str
    starts: int
    avg_finishers: float | None = None


class LocationWeatherYearAgoResponse(BaseModel):
    years: int
    weather: WeatherBriefResponse
    start: WeatherStartRefResponse | None = None


class LocationWeatherResponse(BaseModel):
    slug: str
    name: str
    has_data: bool
    months: list[LocationWeatherMonthResponse] = Field(default_factory=list)
    records: dict[str, WeatherRecordResponse | None] = Field(default_factory=dict)
    latest: WeatherRecordResponse | None = None
    years_ago: list[LocationWeatherYearAgoResponse] = Field(default_factory=list)
    attendance: list[LocationWeatherAttendanceResponse] = Field(default_factory=list)
    starts_with_weather: int = 0


class WeekWeatherLocationResponse(BaseModel):
    location_name: str
    location_slug: str | None = None
    platform_code: str
    weather: WeatherBriefResponse


class WeekWeatherResponse(BaseModel):
    locations_with_weather: int = 0
    temperature_median_c: float | None = None
    rain_locations: int = 0
    coldest: WeekWeatherLocationResponse | None = None
    warmest: WeekWeatherLocationResponse | None = None
    wettest: WeekWeatherLocationResponse | None = None
    windiest: WeekWeatherLocationResponse | None = None
    is_preliminary: bool = False


class UserWeatherRunResponse(BaseModel):
    event_date: str
    platform_code: str
    location_name: str
    location_slug: str | None = None
    weather: WeatherBriefResponse


class UserWeatherStatsResponse(BaseModel):
    runs_with_weather: int = 0
    coldest: UserWeatherRunResponse | None = None
    hottest: UserWeatherRunResponse | None = None
    wettest: UserWeatherRunResponse | None = None
    windiest: UserWeatherRunResponse | None = None
    snowiest: UserWeatherRunResponse | None = None
    rain_runs: int = 0
    frost_runs: int = 0
    heat_runs: int = 0
    snow_runs: int = 0
